### PR TITLE
Charge command refactor

### PR DESCRIPTION
**Pull request by lasrinnil**
## Description

This PR will refactor our ChargeCommand model to no longer so many null able properties. Along with checking when receiving a message that the ChargeCommand within's properties are not null where they are required by business logic to not be. 

We have tested our happy flow with serializing and de-serializing the refactored ChargeCommand.

image

## References

* #116

### CODE REVIEW

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/ChangeOfCharges/Transaction/ChargeCommand.cs [hunk 1]
```diff
@@ -21,32 +21,32 @@ namespace GreenEnergyHub.Charges.Domain.ChangeOfCharges.Transaction
 {
     public class ChargeCommand : CommandBase
     {
-        public MarketDocument? MarketDocument { get; set; }
+#pragma warning disable 8618
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/ChangeOfCharges/Transaction/ChargeCommand.cs:24]: I'd like a comment about what 8618 is and why you use it because that's important and not evident from the code.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/ChangeOfCharges/Transaction/ChargeType.cs [hunk 2]
```diff
@@ -18,17 +18,19 @@ namespace GreenEnergyHub.Charges.Domain.ChangeOfCharges.Transaction
 {
     public class ChargeType
     {
+#pragma warning disable 8618
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/ChangeOfCharges/Transaction/ChargeType.cs:21]: Just thoughts: Can we apply this disabling only once? E.g. using a "local" configuration file? Or by creating a project for DTOs? Or do we actually want this explicit disabling in each class?

**Review by BjarkeMeier — COMMENTED** [review]
I assume that you have tested that the changes doesn't break the deserialization of messages received from the service bus(?)

It's nice if we can get rid of all the nullable props this way. I think the PR should also fix all the - now - superfluous null forgiving operators.

**Comment by lasrinnil** [comment]
> 
> 
> I assume that you have tested that the changes doesn't break the deserialization of messages received from the service bus(?)
> 
> It's nice if we can get rid of all the nullable props this way. I think the PR should also fix all the - now - superfluous null forgiving operators.

I Would like address this specifically in a PR of its own.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/ChangeOfCharges/Transaction/ChargeCommand.cs (hunk 1 above)
Comment (lasrinnil) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/ChangeOfCharges/Transaction/ChargeCommand.cs]: I have put a comment in each file ignored the CA rule now

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/ChangeOfCharges/Transaction/ChargeType.cs (hunk 2 above)
Comment (lasrinnil) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/ChangeOfCharges/Transaction/ChargeType.cs]: We could disable it in a Globalsettings.cs file i guess, but I like that we just ignore it in the specific cases we dont want the compiler to worry about non nullable props not being instantiated

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/GlobalSuppressions.cs [hunk 3]
```diff
@@ -0,0 +1,16 @@
+﻿//---------------------------------------------------------------------
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/GlobalSuppressions.cs]: Should we remove this if it is not used?

Comment (lasrinnil) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/GlobalSuppressions.cs]: Haha, yes it is removed (should not have been checked in, DOH!)

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/ChangeOfCharges/Transaction/ChargeCommand.cs [hunk 4]
```diff
@@ -16,37 +16,39 @@
...
-        public MktActivityRecord? MktActivityRecord { get; set; }
+        public MktActivityRecord MktActivityRecord { get; set; }
 
         /// <summary>
         /// The kind of charge: Fee ("D02") | Subscription ("D01") | Tariff ("D03").
         /// </summary>
-        public string? Type { get; set; }
+        public string Type { get; set; }
 
         [JsonPropertyName("ChargeType_mRID")]
```
Comment (bemwen) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/ChangeOfCharges/Transaction/ChargeCommand.cs:37]: We should try to get rid of these json properties if possible, but I assume that is out of scope of this PR?

Comment (lasrinnil) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/ChangeOfCharges/Transaction/ChargeCommand.cs:37]: We will address in a PR which will rename the model and our Postman models at the same time me and @prtandrup

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/ChangeOfCharges/Transaction/ChargeType.cs [hunk 5]
```diff
@@ -13,19 +13,23 @@
 // limitations under the License.
 
 using System.Text.Json.Serialization;
+#pragma warning disable 8618
 
 namespace GreenEnergyHub.Charges.Domain.ChangeOfCharges.Transaction
 {
+    // Non-nullable member is uninitialized is ignored
+    // Only properties which is allowed to be null is nullable
+    // ChargeType integrity is null checked by ChargeCommandNullChecker
```
Comment (bemwen) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Domain/ChangeOfCharges/Transaction/ChargeType.cs:22]: If I had not heard about this, I do not think I would get that comment is actually about why we are using pragma here (and the other examples). Maybe it would be easier to make the connection if they comments was nearer the pragma line.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Infrastructure/Repositories/ChargeRepositoryTests.cs [hunk 6]
```diff
@@ -199,9 +197,12 @@ private static Charge GetValidCharge()
...
             transaction.Period.AddPoints(new List<Point>
-{
-    new Point { Position = 1, PriceAmount = 1m, Time = SystemClock.Instance.GetCurrentInstant(), },
-});
+                    {
+                    new Point
+                    {
+                        Position = 1, PriceAmount = 1m, Time = SystemClock.Instance.GetCurrentInstant(),
+                    },
+                    });
```
Comment (bemwen) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Tests/Infrastructure/Repositories/ChargeRepositoryTests.cs:205]: NIT: Maybe inline the new point segment for readability

**Review by bemwen — APPROVED** [review]
LGTM, only a few NIT comments

**Review by BjarkeMeier — APPROVED** [review]
